'CREATE TABLE AS' failed with varchar type column

## What happened

When using [`CREATE TABLE AS`](https://prestosql.io/docs/325/sql/create-table-as.html), it throws SQL Error when target table contains column in varchar type:

```SQL
CREATE TABLE IF NOT EXISTS bludb.xxx.summary_table_1
COMMENT 'summary table for test'
AS
SELECT
    last_day_of_week -- varchar
FROM hive.default.fact_table_1
WHERE dt = '2020-02-02'
```

It will throw this error message when running ☝️SQL via prestosql:

```
Query 20200204_020051_00006_8ks7p failed: DB2 SQL Error: SQLCODE=-604, SQLSTATE=42611, SQLERRMC=VARCHAR, DRIVER=4.25.13
```

# Expectation
It should create a new table `summary_table_1` with data from ☝️ query from the Hive table `fact_table_1`.

More information found from prestosql query details page.

stacktrace:

```
io.prestosql.spi.PrestoException: DB2 SQL Error: SQLCODE=-604, SQLSTATE=42611, SQLERRMC=VARCHAR, DRIVER=4.25.13
	at io.prestosql.plugin.jdbc.BaseJdbcClient.beginCreateTable(BaseJdbcClient.java:366)
	at io.prestosql.plugin.jdbc.jmx.StatisticsAwareJdbcClient.lambda$beginCreateTable$16(StatisticsAwareJdbcClient.java:174)
	at io.prestosql.plugin.jdbc.jmx.JdbcApiStats.wrap(JdbcApiStats.java:35)
	at io.prestosql.plugin.jdbc.jmx.StatisticsAwareJdbcClient.beginCreateTable(StatisticsAwareJdbcClient.java:174)
	at io.prestosql.plugin.jdbc.ForwardingJdbcClient.beginCreateTable(ForwardingJdbcClient.java:110)
	at io.prestosql.plugin.jdbc.ForwardingJdbcClient.beginCreateTable(ForwardingJdbcClient.java:110)
	at io.prestosql.plugin.jdbc.JdbcMetadata.beginCreateTable(JdbcMetadata.java:211)
	at io.prestosql.metadata.MetadataManager.beginCreateTable(MetadataManager.java:762)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.createWriterTarget(BeginTableWrite.java:179)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:142)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite$Rewriter.visitTableFinish(BeginTableWrite.java:76)
	at io.prestosql.sql.planner.plan.TableFinishNode.accept(TableFinishNode.java:106)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:84)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:73)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Collections$2.tryAdvance(Collections.java:4747)
	at java.base/java.util.Collections$2.forEachRemaining(Collections.java:4755)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:74)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:38)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.visitPlan(SimplePlanRewriter.java:22)
	at io.prestosql.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.prestosql.sql.planner.plan.OutputNode.accept(OutputNode.java:82)
	at io.prestosql.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:32)
	at io.prestosql.sql.planner.optimizations.BeginTableWrite.optimize(BeginTableWrite.java:73)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:180)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:169)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:164)
	at io.prestosql.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:396)
	at io.prestosql.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:384)
	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:336)
	at io.prestosql.$gen.Presto_325____20200204_013357_2.run(Unknown Source)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:240)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$4(LocalDispatchQuery.java:107)
	at io.prestosql.$gen.Presto_325____20200204_013357_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.ibm.db2.jcc.am.SqlSyntaxErrorException: DB2 SQL Error: SQLCODE=-604, SQLSTATE=42611, SQLERRMC=VARCHAR, DRIVER=4.25.13
	at com.ibm.db2.jcc.am.b6.a(b6.java:810)
	at com.ibm.db2.jcc.am.b6.a(b6.java:66)
	at com.ibm.db2.jcc.am.b6.a(b6.java:140)
	at com.ibm.db2.jcc.am.k3.c(k3.java:2824)
	at com.ibm.db2.jcc.am.k3.d(k3.java:2808)
	at com.ibm.db2.jcc.am.k3.b(k3.java:2168)
	at com.ibm.db2.jcc.t4.ab.k(ab.java:446)
	at com.ibm.db2.jcc.t4.ab.c(ab.java:104)
	at com.ibm.db2.jcc.t4.p.b(p.java:38)
	at com.ibm.db2.jcc.t4.av.h(av.java:124)
	at com.ibm.db2.jcc.am.k3.ak(k3.java:2163)
	at com.ibm.db2.jcc.am.k3.a(k3.java:3366)
	at com.ibm.db2.jcc.am.k3.e(k3.java:1131)
	at com.ibm.db2.jcc.am.k3.execute(k3.java:1110)
	at io.prestosql.plugin.jdbc.BaseJdbcClient.execute(BaseJdbcClient.java:779)
	at io.prestosql.plugin.jdbc.BaseJdbcClient.createTable(BaseJdbcClient.java:406)
	at io.prestosql.plugin.jdbc.BaseJdbcClient.beginCreateTable(BaseJdbcClient.java:363)
	... 41 more
```

Fixing this issue under branch `varchar`: https://github.com/shawnzhu/presto-db2/tree/varchar

Got new error message after fixing mapping unbounded varchar to `varchar(32672)`:

```
io.prestosql.spi.PrestoException: DB2 SQL Error: SQLCODE=-104, SQLSTATE=42601, SQLERRMC=ALTER TABLE HCW88054.TMP_PRESTO_1;BEGIN-OF-STATEMENT;<rename_stogroup>, DRIVER=4.25.13
	at io.prestosql.plugin.jdbc.BaseJdbcClient.renameTable(BaseJdbcClient.java:523)
	at io.prestosql.plugin.jdbc.BaseJdbcClient.commitCreateTable(BaseJdbcClient.java:493)
	at io.prestosql.plugin.jdbc.jmx.StatisticsAwareJdbcClient.lambda$commitCreateTable$17(StatisticsAwareJdbcClient.java:180)
	at io.prestosql.plugin.jdbc.jmx.JdbcApiStats.wrap(JdbcApiStats.java:47)
	at io.prestosql.plugin.jdbc.jmx.StatisticsAwareJdbcClient.commitCreateTable(StatisticsAwareJdbcClient.java:180)
	at io.prestosql.plugin.jdbc.ForwardingJdbcClient.commitCreateTable(ForwardingJdbcClient.java:116)
	at io.prestosql.plugin.jdbc.ForwardingJdbcClient.commitCreateTable(ForwardingJdbcClient.java:116)
	at io.prestosql.plugin.jdbc.JdbcMetadata.finishCreateTable(JdbcMetadata.java:226)
	at io.prestosql.metadata.MetadataManager.finishCreateTable(MetadataManager.java:771)
	at io.prestosql.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$3(LocalExecutionPlanner.java:2852)
	at io.prestosql.operator.TableFinishOperator.getOutput(TableFinishOperator.java:297)
	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
	at io.prestosql.operator.Driver.processFor(Driver.java:276)
	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.prestosql.$gen.Presto_325____20200205_024616_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.ibm.db2.jcc.am.SqlSyntaxErrorException: DB2 SQL Error: SQLCODE=-104, SQLSTATE=42601, SQLERRMC=ALTER TABLE HCW88054.TMP_PRESTO_1;BEGIN-OF-STATEMENT;<rename_stogroup>, DRIVER=4.25.13
	at com.ibm.db2.jcc.am.b6.a(b6.java:810)
	at com.ibm.db2.jcc.am.b6.a(b6.java:66)
	at com.ibm.db2.jcc.am.b6.a(b6.java:140)
	at com.ibm.db2.jcc.am.k3.c(k3.java:2824)
	at com.ibm.db2.jcc.am.k3.d(k3.java:2808)
	at com.ibm.db2.jcc.am.k3.b(k3.java:2168)
	at com.ibm.db2.jcc.t4.ab.k(ab.java:446)
	at com.ibm.db2.jcc.t4.ab.c(ab.java:104)
	at com.ibm.db2.jcc.t4.p.b(p.java:38)
	at com.ibm.db2.jcc.t4.av.h(av.java:124)
	at com.ibm.db2.jcc.am.k3.ak(k3.java:2163)
	at com.ibm.db2.jcc.am.k3.a(k3.java:3366)
	at com.ibm.db2.jcc.am.k3.e(k3.java:1131)
	at com.ibm.db2.jcc.am.k3.execute(k3.java:1110)
	at io.prestosql.plugin.jdbc.BaseJdbcClient.execute(BaseJdbcClient.java:779)
	at io.prestosql.plugin.jdbc.BaseJdbcClient.renameTable(BaseJdbcClient.java:520)
	... 21 more
```

--- a/src/main/java/io/prestosql/plugin/db2/DB2Client.java
+++ b/src/main/java/io/prestosql/plugin/db2/DB2Client.java
@@ -19,11 +19,18 @@ import io.prestosql.plugin.jdbc.ConnectionFactory;
 import io.prestosql.plugin.jdbc.JdbcIdentity;
 import io.prestosql.plugin.jdbc.StatsCollecting;
 import io.prestosql.plugin.jdbc.JdbcSplit;
+import io.prestosql.plugin.jdbc.WriteMapping;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.VarcharType;
 
 import javax.inject.Inject;
 
 import java.sql.Connection;
 import java.sql.SQLException;
+
+import static io.prestosql.plugin.jdbc.StandardColumnMappings.varcharWriteFunction;
+import static io.prestosql.spi.type.Varchars.isVarcharType;
 
 public class DB2Client
         extends BaseJdbcClient
@@ -52,5 +59,33 @@ public class DB2Client
             throw e;
         }
         return connection;
+    }
+    
+    /**
+     * To map data types when generating SQL.
+     */
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+    	if (isVarcharType(type)) {
+            VarcharType varcharType = (VarcharType) type;
+            String dataType;
+            if (varcharType.isUnbounded()) {
+            	dataType = "VARCHAR(32672)";
+            }
+            else if (varcharType.getBoundedLength() > 32672) {
+                dataType = "CLOB(" + varcharType.getBoundedLength() + ")";
+            }
+            else if (varcharType.getBoundedLength() < 32672) {
+                dataType = "VARCHAR(" + varcharType.getBoundedLength() + ")";
+            }
+            else {
+            	// The maximum length of VARCHAR is 32672
+                dataType = "VARCHAR(32672)";
+            }
+            return WriteMapping.sliceMapping(dataType, varcharWriteFunction());
+        }
+    	
+    	return super.toWriteMapping(session, type);
     }
 }

--- a/src/main/java/io/prestosql/plugin/db2/DB2Client.java
+++ b/src/main/java/io/prestosql/plugin/db2/DB2Client.java
@@ -17,10 +17,12 @@ import io.prestosql.plugin.jdbc.BaseJdbcClient;
 import io.prestosql.plugin.jdbc.BaseJdbcConfig;
 import io.prestosql.plugin.jdbc.ConnectionFactory;
 import io.prestosql.plugin.jdbc.JdbcIdentity;
-import io.prestosql.plugin.jdbc.StatsCollecting;
 import io.prestosql.plugin.jdbc.JdbcSplit;
+import io.prestosql.plugin.jdbc.StatsCollecting;
 import io.prestosql.plugin.jdbc.WriteMapping;
+import io.prestosql.spi.PrestoException;
 import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.connector.SchemaTableName;
 import io.prestosql.spi.type.Type;
 import io.prestosql.spi.type.VarcharType;
 
@@ -29,8 +31,10 @@ import javax.inject.Inject;
 import java.sql.Connection;
 import java.sql.SQLException;
 
+import static io.prestosql.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
 import static io.prestosql.plugin.jdbc.StandardColumnMappings.varcharWriteFunction;
 import static io.prestosql.spi.type.Varchars.isVarcharType;
+import static java.lang.String.format;
 
 public class DB2Client
         extends BaseJdbcClient
@@ -86,6 +90,24 @@ public class DB2Client
             return WriteMapping.sliceMapping(dataType, varcharWriteFunction());
         }
     	
-    	return super.toWriteMapping(session, type);
+        return super.toWriteMapping(session, type);
+    }
+
+    @Override
+    protected void renameTable(JdbcIdentity identity, String catalogName, String schemaName, String tableName, SchemaTableName newTable)
+    {
+    	// TODO figure if it supports changing schema while renaming table
+
+        String sql = format(
+                "RENAME TABLE %s TO %s",
+                quoted(catalogName, schemaName, tableName),
+                quoted(newTable.getTableName()));
+
+        try (Connection connection = connectionFactory.openConnection(identity)) {
+            execute(connection, sql);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
     }
 }

